File: 0428-Serialize and Deserialize N-ary Tree/0428-Serialize and Deserialize N-ary Tree.cpp
/*
// Definition for a Node.
class Node {
public:
    int val = NULL;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(Node* root) {
        if (root == nullptr) {
            return "[]";
        }
        string result = "[" + to_string(root->val);
        for (Node* child : root->children) {
            result += serialize(child);
        }
        result += "]";

        return result;
    }

    // Decodes your encoded data to tree.
    Node* deserialize(string data) {
        if (data == "[]") {
            return nullptr;
        }
        
        int i = 1, num = 0, n = data.size();
        while (i < n - 1 && isdigit(data[i])) {
            num = num * 10 + (data[i++] - '0');
        }
        Node* node = new Node(num, {});
        stack<Node*> St;
        while (i < n - 1) {
            if (data[i] == '[') {
                St.push(node);
                ++i;
                num = 0;
                while (i < n - 1 && isdigit(data[i])) {
                    num = num * 10 + (data[i++] - '0');
                }
                node = new Node(num, {});
            }
            else if (data[i] == ']') {
                ++i;
                St.top()->children.push_back(node);
                node = St.top();
                St.pop();              
            }
        }
        
        return node;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));

File: 0401-0500/0428-Serialize and Deserialize N-ary Tree/0428-Serialize and Deserialize N-ary Tree.py
# Definition for a Node.
class Node(object):
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children

class Codec:

    def serialize(self, root: 'Node') -> str:
        """Encodes a tree to a single string.
        
        :type root: Node
        :rtype: str
        """
        if root is None:
            return '[]'
        result = ['[', str(root.val)]
        for child in root.children:
            result.append(self.serialize(child))
        result.append(']')
        return ''.join(result)

    def deserialize(self, data: str) -> 'Node':
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: Node
        """
        if data == '[]':
            return None
        i = 1
        num = 0
        while i < len(data) - 1 and data[i].isdigit():
            num = num * 10 + int(data[i])
            i += 1
        node = Node(num, [])
        St = []
        while i < len(data) - 1:
            if data[i] == '[':
                St.append(node)
                i += 1
                num = 0
                while i < len(data) - 1 and data[i].isdigit():
                    num = num * 10 + int(data[i])
                    i += 1
                node = Node(num, [])
            elif data[i] == ']':
                i += 1
                St[-1].children.append(node)
                node = St.pop()
                
        return node

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))
